tests/bsim/bluetooth/host/att/long_read/testlib/att_read.c

```diff
@@ -0,0 +1,390 @@
+/* Copyright (c) 2023 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/gatt.h>
+#include <zephyr/bluetooth/att.h>
+#include <zephyr/bluetooth/l2cap.h>
+#include <zephyr/bluetooth/uuid.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/__assert.h>
+
+#include "att_read.h"
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(att_read, LOG_LEVEL_DBG);
+
+struct bt_testlib_att_read_closure {
+	uint8_t att_err;
+	struct bt_conn *conn;
+	struct bt_gatt_read_params params;
+	uint16_t *result_size;
+	uint16_t *result_handle;
+	struct net_buf_simple *result_data;
+	struct k_mutex lock;
+	struct k_condvar done;
+	uint16_t *att_mtu;
+	bool long_read;
+};
+
+static inline uint8_t att_read_cb(struct bt_conn *conn, uint8_t att_err,
+				  struct bt_gatt_read_params *params, const void *read_data,
+				  uint16_t read_len)
+{
+	struct bt_testlib_att_read_closure *ctx =
+		CONTAINER_OF(params, struct bt_testlib_att_read_closure, params);
+
+	k_mutex_lock(&ctx->lock, K_FOREVER);
+
+	if (read_data == NULL) {
+		__ASSERT_NO_MSG(ctx->long_read);
+		k_condvar_signal(&ctx->done);
+		k_mutex_unlock(&ctx->lock);
+		return BT_GATT_ITER_STOP;
+	}
+
+	ctx->att_err = att_err;
+
+	if (!att_err && ctx->result_handle) {
+		*ctx->result_handle = params->by_uuid.start_handle;
+	}
+
+	if (!att_err && ctx->result_size) {
+		LOG_DBG("Adding %u bytes to result", read_len);
+		*ctx->result_size += read_len;
+		if (*ctx->result_size > 512) {
+			LOG_ERR("result_size > 512");
+		}
+	}
+
+	if (!att_err && ctx->result_data) {
+		uint16_t result_data_size =
+			MIN(read_len, net_buf_simple_tailroom(ctx->result_data));
+		net_buf_simple_add_mem(ctx->result_data, read_data, result_data_size);
+	}
+
+	if (!att_err && ctx->att_mtu) {
+		*ctx->att_mtu = params->_att_mtu;
+	}
+
+	if (ctx->long_read) {
+		/* Don't signal `&ctx->done` */
+		k_mutex_unlock(&ctx->lock);
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	k_condvar_signal(&ctx->done);
+	k_mutex_unlock(&ctx->lock);
+	return BT_GATT_ITER_STOP;
+}
+
+static inline int bt_testlib_sync_bt_gatt_read(struct bt_testlib_att_read_closure *ctx)
+{
+	int api_err;
+
+	ctx->params.func = att_read_cb;
+
+	k_mutex_init(&ctx->lock);
+	k_condvar_init(&ctx->done);
+
+	k_mutex_lock(&ctx->lock, K_FOREVER);
+
+	api_err = bt_gatt_read(ctx->conn, &ctx->params);
+
+	if (!api_err) {
+		k_condvar_wait(&ctx->done, &ctx->lock, K_FOREVER);
+	}
+
+	k_mutex_unlock(&ctx->lock);
+
+	if (api_err) {
+		__ASSERT_NO_MSG(api_err < 0);
+		return api_err;
+	}
+
+	__ASSERT_NO_MSG(ctx->att_err >= 0);
```

Comment:
This is an actual programmer's assertion. It's just for show actually since `att_err` is unsigned, so this assert can never fire.

I also avoid bsim as a dependency for testlib modules that don't need it. This module should run perfectly fine on hardware.

Reply:
Alright, but if you want to rely on `__ASSERT` for test validation, then you may want to consider a `BUILD_ASSERT` to ensure that the test isn't run without ASSERT enabled.

Reply:
Can I consider this conversation resolved?

Reply:
I'd still prefer either failing with the BSIM API, or ensure that ASSERT is enabled. 

Having a test library that can run without reporting errors is a bit odd IMO. 

But since this is a corner case (i.e. someone will have to disable asserts for this to be a problem) I guess it's fine.